docs/v3/core/index.rst

```diff
@@ -1599,46 +1466,207 @@ Storage transformers may be stacked to combine different functionalities:
+
+.. _extension-definition:
+
+Extension definition
+--------------------
+
+.. _extension-definition-object:
+
+Objects
+^^^^^^^
+
+In `metadata documents`_, extensions can be encoded either as objects or as
+short-hand names.
+
+If using an object definition, the member ``name``
+MUST be a plain string which conforms to :ref:`extension name <extension-naming>`.
+Optionally, the member ``configuration`` MAY be present but if so MUST be
+an object.
+
+For example::
+
+    {
+        "name": "<name>",        # conformant name
+        "configuration": { ... } # optional object
+    }
+
+.. _extension-definition-short-hand-name:
+
+Short-hand names
+^^^^^^^^^^^^^^^^
+
+Instead of extension objects, short-hand names MAY be used if no
+configuration metadata is required. They are equivalent to extension
+objects with just a `name` key.
+
+.. _extension-definition-must-understand:
+
+`must_understand`
+^^^^^^^^^^^^^^^^^
+
+An extension object is interpreted to have an implicit field `must_understand` set to
+`True`, unless otherwise stated. An extension object MAY explicitly set `must_understand=False` if
+implementations can ignore its presence.
+
+An implementation MUST fail to open Zarr groups or arrays if any
+metadata fields are present which (a) the
+implementation does not recognize and (b) are not explicitly
+set to ``"must_understand": false``.
+
+`must_understand=False` is not supported for the following extension points:
+data type, chunk grid, and chunk key encoding.
+
+Use of `must_understand=False` to add top-level keys is discouraged in favor
+of the explicit use of :ref:`extension-points`.
+
+.. _extension-naming:
+
+Extension naming
+----------------
+
+The `name` field of an extension is an identifier that has been registered
+prior to release in any implementation within the `zarr-extensions`_ Github
+repository, where extensions and their specification are listed. The Zarr
+Steering Council or by delegation a maintainer team reserves the right to
+refuse name assignment at its own discretion.
+
+.. _extension-naming-registered-names:
+
+Registered names consist of a single string that is unique within the Zarr ecosystem.
+Registered names are intended for well-known extensions aimed at broad adoption and maximum interoperability.
+Registered names are unique and immutable.
+
+Registered names MUST start with one lower case letter a-z and then be followed
+by only lower case letters a-z, numerals 0-9, underscores, dots and dashes.
+
+- **Accepted regex:** ``^[a-z][a-z0-9-_.]+$``
+- **Valid examples:**
+    - ``zstd``
+    - ``numcodecs.adler32``
+- **Invalid examples:**
+    - ``foo/bar``
+    - ``foo:bar``
+
+.. note::
+  In previous versions of the v3 spec, the name of an extension was required
+  to be a URI. That is now discouraged for new extensions, though, for
+  backwards compatibility with existing extensions, URIs names are still
+  permitted.
```

Comment:
Does this note need to be updated or perhaps removed to better reflect the current position?

Reply:
I.e., IIUC, registered extensions will not have a URI name, and registration is encouraged. But for now we'd like to leave open the option that URIs can also be used as extension names to provide a decentralised mechanism to avoid extension name clashes. So saying "URIs are discouraged for new extensions" in this note is potentially confusing?

Reply:
I think this is fundamentally what over half the conversation here was about. I agree that the current state is confusing, or at least ambiguous, but that is the current state of the spec and I'd leave this to a follow on PR.